File: tests/metadata.py
"""Validate public contracts and nested action wiring without network access."""

from pathlib import Path
import re
import yaml

root = Path(__file__).resolve().parent.parent
public = ["setup-nix", "setup-devenv", "setup-trivy", "setup-cache", "free-disk-space"]
reviewed = {
    "actions/cache": "55cc8345863c7cc4c66a329aec7e433d2d1c52a9",
    "runs-on/cache": "88d90644011a3a9957fd141a106f5a94f9794203",
    "cachix/install-nix-action": "13d8dd58da0234aa297dedd986986ccb8e7f3e24",
    "cachix/cachix-action": "38b082610b782e7e93e209c35fd730d399dee866",
    "actions/checkout": "3d3c42e5aac5ba805825da76410c181273ba90b1",
}


def load(name):
    return yaml.safe_load((root / name / "action.yml").read_text())


for name in public:
    action = load(name)
    assert action["runs"]["using"] == "composite", name
    assert (root / name / "README.md").is_file(), name

for file in [*(root / "internal").rglob("action.yml"), *[root / name / "action.yml" for name in public]]:
    action = yaml.safe_load(file.read_text())
    if action["runs"]["using"] == "node24":
        assert (file.parent / action["runs"]["main"]).is_file()
        continue
    for step in action["runs"]["steps"]:
        if "run" in step:
            assert step["shell"] == "bash"
            assert "${{" not in step["run"], file
        if "uses" in step:
            reference = step["uses"]
            if reference.startswith("$/"):
                target = root / reference[2:] / "action.yml"
                assert target.is_file(), target
                accepted = yaml.safe_load(target.read_text()).get("inputs", {})
                assert set(step.get("with", {})) <= set(accepted), reference
            else:
                owner, sha = reference.split("@")
                assert reviewed[owner] == sha, reference
        for value in re.findall(r"inputs\.([a-z0-9-]+)", str(step)):
            assert value in action.get("inputs", {}), (file, value)

for name in ["setup-cache", "setup-devenv"]:
    refs = [s.get("uses") for s in load(name)["runs"]["steps"]]
    assert "$/setup-nix" in refs
    assert "$/free-disk-space" not in refs
assert all("cache" not in s.get("uses", "") for s in load("setup-devenv")["runs"]["steps"])
assert all("cache" not in s.get("uses", "") for s in load("setup-trivy")["runs"]["steps"])
# Composite post hooks evaluate outputs without the main steps' JSON context.
assert "fromJSON(" not in str(load("setup-cache")["outputs"])
adapter = load("internal/cache")["runs"]["steps"]
github, s3 = adapter[:2]
assert github["if"] == "inputs.backend == 'github'"
assert s3["if"] == "inputs.backend == 's3'"
assert s3["continue-on-error"] is True
assert s3["env"]["RUNS_ON_RUNNER_NAME"] == ""
assert "AWS_ACCESS_KEY_ID" not in github.get("env", {})
assert "save-always" not in str(adapter)
ci = yaml.safe_load((root / ".github/workflows/ci.yml").read_text())
assert ci["jobs"]["cache-warm"]["needs"] == "cache-cold"
assert ci["jobs"]["flakes"]["strategy"]["matrix"]["shell"] == ["default", "named"]
for name, job in ci["jobs"].items():
    if name == "self-hosted":
        assert job["runs-on"]["group"] == "enterprise/tars-cloud"
        assert all(s.get("uses") != "./free-disk-space" for s in job["steps"])
        continue
    assert job["strategy"]["matrix"]["runner"] == ["ubuntu-24.04", "ubuntu-24.04-arm"]
print("Action metadata, immutable pins, isolation and CI lifecycle wiring passed.")

File: crates/tact/src/checks/metadata.rs
use anyhow::{Context, Result, ensure};
use serde_json::{Value, json};
use std::fs;
use std::path::Path;

const PINS: [(&str, &str); 5] = [
    ("actions/cache", "55cc8345863c7cc4c66a329aec7e433d2d1c52a9"),
    ("runs-on/cache", "88d90644011a3a9957fd141a106f5a94f9794203"),
    (
        "cachix/install-nix-action",
        "13d8dd58da0234aa297dedd986986ccb8e7f3e24",
    ),
    (
        "cachix/cachix-action",
        "38b082610b782e7e93e209c35fd730d399dee866",
    ),
    (
        "actions/checkout",
        "3d3c42e5aac5ba805825da76410c181273ba90b1",
    ),
];
fn load(path: impl AsRef<Path>) -> Result<Value> {
    Ok(serde_norway::from_str(&fs::read_to_string(path)?)?)
}

pub(super) fn adapter(root: &Path) -> Result<()> {
    let a = load(root.join("composite/setup-cache/scripts/cache/action.yml"))?;
    let steps = &a["runs"]["steps"];
    ensure!(
        steps[0]["if"] == "inputs.backend == 'github'",
        "GitHub backend must be exclusive"
    );
    ensure!(
        steps[1]["if"] == "inputs.backend == 's3'",
        "S3 backend must be exclusive"
    );
    ensure!(
        steps[1]["continue-on-error"] == true,
        "S3 must remain nonfatal"
    );
    ensure!(
        steps[1]["env"]["RUNS_ON_RUNNER_NAME"] == "",
        "ambient RunsOn routing must be disabled"
    );
    ensure!(
        steps[0]["env"]["AWS_ACCESS_KEY_ID"].is_null(),
        "S3 credentials leaked into GitHub backend"
    );
    ensure!(
        !steps.to_string().contains("save-always"),
        "cache saves must be success-only"
    );
    Ok(())
}

pub(super) fn run(root: &Path) -> Result<()> {
    let suites = crate::manifest::discover(root, None)?;
    for suite in &suites {
        let name = suite.action.as_str();
        let public_name = name
            .strip_prefix("composite/")
            .context("action collection")?;
        let owner = format!(
            "composite/{}",
            public_name.split('/').next().context("action owner")?
        );
        ensure!(
            suite
                .manifest
                .sources
                .iter()
                .all(|source| source.starts_with(&format!("{owner}/"))),
            "test sources must belong to their public action: {name}"
        );
        let a = load(root.join(name).join("action.yml"))?;
        if !public_name.contains('/') {
            ensure!(
                a["runs"]["using"] == "composite" && root.join(name).join("README.md").is_file(),
                "public contract: {name}"
            );
        }
        if a["runs"]["using"] == "node24" {
            ensure!(
                root.join(name)
                    .join(a["runs"]["main"].as_str().context("Node entrypoint")?)
                    .is_file(),
                "missing Node entrypoint"
            );
            continue;
        }
        let steps = a["runs"]["steps"].as_array().context("composite steps")?;
        for step in steps {
            if let Some(run) = step["run"].as_str() {
                ensure!(
                    step["shell"] == "bash" && !run.contains("${{") && !run.contains("../"),
                    "unsafe inline shell: {name}"
                );
            }
            if let Some(reference) = step["uses"].as_str() {
                if let Some(local) = reference.strip_prefix("$/") {
                    ensure!(
                        local
                            .strip_prefix("composite/")
                            .is_some_and(|path| !path.contains('/'))
                            || local.starts_with(&format!("{owner}/scripts/")),
                        "private helper must belong to its caller: {name} -> {local}"
                    );
                    let target = load(root.join(local).join("action.yml"))?;
                    if let Some(inputs) = step["with"].as_object() {
                        for key in inputs.keys() {
                            ensure!(
                                target["inputs"].get(key).is_some(),
                                "unknown input {local}/{key}"
                            );
                        }
                    }
                } else {
                    let (owner, sha) = reference
                        .split_once('@')
                        .context("unpinned upstream action")?;
                    ensure!(PINS.contains(&(owner, sha)), "unreviewed pin: {reference}");
                }
            }
            let text = step.to_string();
            for rest in text.split("inputs.").skip(1) {
                let key: String = rest
                    .chars()
                    .take_while(|c| c.is_ascii_alphanumeric() || *c == '-')
                    .collect();
                ensure!(
                    a["inputs"].get(&key).is_some(),
                    "unknown {name} input: {key}"
                );
            }
        }
        if ["composite/setup-cache", "composite/setup-devenv"].contains(&name) {
            ensure!(
                steps.iter().any(|s| s["uses"] == "$/composite/setup-nix"),
                "missing same-revision Nix prerequisite"
            );
            ensure!(
                !steps
                    .iter()
                    .any(|s| s["uses"] == "$/composite/free-disk-space"),
                "implicit cleanup"
            );
        }
        if ["composite/setup-devenv", "composite/setup-trivy"].contains(&name) {
            ensure!(
                !steps
                    .iter()
                    .any(|s| s["uses"].as_str().is_some_and(|s| s.contains("cache"))),
                "implicit cache setup"
            );
        }
        if name == "composite/setup-cache" {
            ensure!(
                !a["outputs"].to_string().contains("fromJSON("),
                "post hooks cannot reevaluate JSON outputs"
            );
        }
    }
    adapter(root)?;
    let ci = load(root.join(".github/workflows/ci.yml"))?;
    ensure!(
        ci["jobs"]["cache-warm"]["needs"] == "cache-cold",
        "cold/warm lifecycle dependency"
    );
    ensure!(
        ci["jobs"]["flakes"]["strategy"]["matrix"]["shell"] == json!(["default", "named"]),
        "flake shell matrix"
    );
    ensure!(
        ci["jobs"]["s3-cache-warm"]["needs"] == "s3-cache-cold",
        "S3 post-save must finish before warm restoration"
    );
    for (name, job) in ci["jobs"].as_object().context("CI jobs")? {
        if name == "self-hosted" || name.starts_with("s3-cache-") {
            ensure!(
                job["runs-on"]["group"] == "enterprise/tars-cloud",
                "enterprise runner group"
            );
            ensure!(
                !job["steps"]
                    .as_array()
                    .context("steps")?
                    .iter()
                    .any(|s| s["uses"] == "./composite/free-disk-space"),
                "persistent runner cleanup"
            );
        } else {
            ensure!(
                job["strategy"]["matrix"]["runner"] == json!(["ubuntu-24.04", "ubuntu-24.04-arm"]),
                "native architecture matrix: {name}"
            );
        }
    }
    Ok(())
}
